[13.0] Improve customer validation, add manage users service

simahawk opened the pull request.

simahawk committed:
Services ctx: introduce invader_partner_user

As the shop partner might be different from current partner user
the related invader partner can be different.

From now on, rely on invader_partner_user to access info
on the current logged in user.
Rely on invader_partner to access info on current shop partner
(normally the main account).

simahawk committed:
shopinvader: refactor partner validation

Customers (bindings) now have 3 states: pending, active, inactive.
Addresses keep the flag only as they are not bound to a backend.
For both of them a wizard is provided to properly change their status
and get users notified.

simahawk committed:
s_customer_multi_user: add enpoints to manage users

simahawk commented:
I included #807 because it's getting merged and it will conflict soon here.

simahawk committed:
s_customer_multi_user: hide fields when not needed

simahawk committed:
s_customer_multi_user: fix tests

simahawk committed:
shopinvader: partner form quick edit form for invader partner

simahawk committed:
Make pre-commit happy

simahawk committed:
s_customer_multi_user: only owners and admin can edit addresses

mmequignon commented on shopinvader_customer_multi_user/components/access_info.py:
```diff
@@ -0,0 +1,30 @@
+        # - owner
+        # - admin account
+        if self.partner_user is not None:
+            can_crud = (
+                self.is_owner(address_id)
+                or self.invader_partner_user.is_admin_account
+            )
+            if can_crud:
+                info.update({"read": True, "update": True, "delete": True})
+            else:
+                info.update({"read": True, "update": False, "delete": False})
```
Only cosmetic:
I think this variable is named wrong, since it's not about creating or reading here.
I would personally name it `can_edit` instead.

mmequignon commented on shopinvader/components/res_partner_event_listener.py:
```diff
@@ -0,0 +1,32 @@
+
+    def _notify_address_validate(self, partner, backends):
+        """Notify address' owner."""
+        notif_type = "address_validated"
```
Should it be global?

mmequignon commented on shopinvader/components/shopinvader_partner_event_listener.py:
```diff
@@ -0,0 +1,34 @@
+        partner = invader_partner.record_id
+        # TODO: handle other states?
+        if invader_partner.state == "active":
+            notif_type = "customer_validated"
```
Same here

mmequignon commented on shopinvader/models/shopinvader_partner.py:
```diff
@@ -7,6 +7,11 @@
+STATE_ACTIVE = "active"
+STATE_INACTIVE = "inactive"
+STATE_PENDING = "pending"
+ALL_STATES = (STATE_ACTIVE, STATE_INACTIVE, STATE_PENDING)
```
Unused?

simahawk replied:
no

mmequignon commented on shopinvader/models/shopinvader_partner.py:
```diff
@@ -185,3 +185,21 @@ def action_shopinvader_validate(self):
+        )
+        view = self.env.ref(form_xid)
+        return {
+            "name": _("Edit %s") % self.name,
```
I think `.format()` is recommended for python3

mmequignon reviewed (approved):
Mostly technical review, LG

HviorForgeFlow commented:
I was testing this functionality, something is not working as expected until last push-force.

Here the error, prompted during last cart checkout step.
```
2021-04-22 15:08:42,155 81 ERROR test odoo.addons.base_rest.http: RESTFULL call to url http://odoo:8069/shopinvader/payment_manual/add_payment with method POST and params {'target': 'current_cart', 'payment_mode_id': '5'} raise the following error 500 Internal Server Error: 'WorkContext' object has no attribute 'invader_partner' 
Traceback (most recent call last):
  File "/opt/odoo/auto/addons/base_rest/http.py", line 182, in _handle_exception
    return super(HttpRestRequest, self)._handle_exception(exception)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 750, in _handle_exception
    return super(HttpRequest, self)._handle_exception(exception)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/opt/odoo/custom/src/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 234, in _dispatch
    result = request.dispatch()
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 809, in dispatch
    r = self._call_function(**self.params)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 350, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 915, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "<string>", line 6, in post_add_payment
  File "/opt/odoo/auto/addons/base_rest/controllers/main.py", line 144, in _process_method
    result = service.dispatch(method_name, *args, params=params)
  File "/opt/odoo/auto/addons/shopinvader/services/service.py", line 247, in dispatch
    res = super().dispatch(method_name, *args, params=params)
  File "/opt/odoo/auto/addons/base_rest/components/service.py", line 158, in dispatch
    res = method(*args, **secure_params)
  File "/opt/odoo/auto/addons/base_rest/restapi.py", line 61, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/auto/addons/invader_payment_manual/services/payment_manual.py", line 54, in add_payment
    transaction.write({"state": "pending"})
  File "/opt/odoo/auto/addons/invader_payment/models/payment_transaction.py", line 45, in write
    self._notify_state_changed_event()
  File "/opt/odoo/auto/addons/invader_payment/models/payment_transaction.py", line 60, in _notify_state_changed_event
    payable._event(event_name).notify(payable, record)
  File "/opt/odoo/auto/addons/component_event/components/event.py", line 191, in notify
    event(*args, **kwargs)
  File "/opt/odoo/auto/addons/shopinvader_payment_manual/components/payment_transaction_event_listerner.py", line 11, in on_payment_transaction_pending
    self._confirm_and_invalidate_session(sale_order)
  File "/opt/odoo/auto/addons/shopinvader_payment/components/payment_transaction_event_listerner.py", line 35, in _confirm_and_invalidate_session
    res = work.component(usage="cart")._to_json(sale_order)
  File "/opt/odoo/auto/addons/shopinvader/services/cart.py", line 428, in _to_json
    res = super(CartService, self)._to_json(cart)[0]
  File "/opt/odoo/auto/addons/shopinvader/services/abstract_sale.py", line 132, in _to_json
    res.append(self._convert_one_sale(sale))
  File "/opt/odoo/auto/addons/shopinvader_payment/services/cart.py", line 20, in _convert_one_sale
    values = super()._convert_one_sale(sale)
  File "/opt/odoo/auto/addons/shopinvader_cart_expiry/services/abstract_sale.py", line 12, in _convert_one_sale
    res = super(AbstractSaleService, self)._convert_one_sale(sale)
  File "/opt/odoo/auto/addons/shopinvader/services/abstract_sale.py", line 30, in _convert_one_sale
    "amount": self._convert_amount(sale),
  File "/opt/odoo/auto/addons/shopinvader_delivery_carrier/services/abstract_sale.py", line 41, in _convert_amount
    shipping_amounts = self._convert_shipping(sale).get("amount", {})
  File "/opt/odoo/auto/addons/shopinvader_delivery_carrier/services/abstract_sale.py", line 12, in _convert_shipping
    res = super(AbstractSaleService, self)._convert_shipping(cart)
  File "/opt/odoo/auto/addons/shopinvader/services/abstract_sale.py", line 103, in _convert_shipping
    "address": address_service._to_json(sale.partner_shipping_id)[
  File "/opt/odoo/auto/addons/shopinvader/services/address.py", line 205, in _to_json
    item["access"] = self.access_info.for_address(item["id"])
  File "/opt/odoo/auto/addons/shopinvader/services/partner_mixin.py", line 29, in access_info
    invader_partner=self.invader_partner,
  File "/opt/odoo/auto/addons/shopinvader/services/service.py", line 35, in invader_partner
    return self.work.invader_partner
AttributeError: 'WorkContext' object has no attribute 'invader_partner'
```
I couldn't find why :cry: 

Great feature! :smiley: I will keep reviewing

simahawk commented:
@HviorForgeFlow sorry but for payments you need this https://github.com/shopinvader/odoo-shopinvader-payment/pull/41 (I forgot to mention it here)

Thanks for reviewing! We are using this in prod since a while now and we have no problem. If I get at least another review from real world I'll cleanup and get ready for merge.

simahawk committed:
s_locomotive: prevent undesired exports on delete

When a record is deleted some fields might get recomputed and cause writes.
When it happens, you can have an export before deletion.
As the record gets deleted right after this is useless.
`on_record_unlink` has no guard for `connector_no_export`
hence it won't be skipped.

simahawk commented:
@HviorForgeFlow would you have time for a review? I'd like to get this merged as we are using it on prod since a while.
It brings some improvements to the core and as it has an upgrade step it's bit annoying to update it every time I rebase :sweat_smile:

HviorForgeFlow commented on shopinvader_customer_multi_user/services/user_management.py:
```diff
@@ -0,0 +1,201 @@
+        params["parent_id"] = parent_id or self.partner_user.id
+        return params
+
+    def _to_json(self, records, **kw):
+        return records.jsonify(self._json_parser(), **kw)
```
There is something wrong with keywords found here, as jsonify is not intend to find any kw.
```2021-05-05 13:40:29,924 1 ERROR devel odoo.addons.base_rest.http: RESTFULL call to url http://odoo:8069/shopinvader/users?page=1&per_page=10 with method GET and params {'page': '1', 'per_page': '10'} raise the following error 500 Internal Server Error: jsonify() got an unexpected keyword argument 'page' 
Traceback (most recent call last):
  File "/opt/odoo/auto/addons/base_rest/http.py", line 184, in _handle_exception
    return super(HttpRestRequest, self)._handle_exception(exception)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 750, in _handle_exception
    return super(HttpRequest, self)._handle_exception(exception)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/opt/odoo/custom/src/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 234, in _dispatch
    result = request.dispatch()
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 809, in dispatch
    r = self._call_function(**self.params)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 350, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 915, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "<string>", line 6, in get_search
  File "/opt/odoo/auto/addons/base_rest/controllers/main.py", line 144, in _process_method
    result = service.dispatch(method_name, *args, params=params)
  File "/opt/odoo/auto/addons/shopinvader/services/service.py", line 247, in dispatch
    res = super().dispatch(method_name, *args, params=params)
  File "/opt/odoo/auto/addons/base_rest/components/service.py", line 158, in dispatch
    res = method(*args, **secure_params)
  File "/opt/odoo/auto/addons/base_rest/restapi.py", line 61, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/auto/addons/shopinvader_customer_multi_user/services/user_management.py", line 39, in wrapped
    return func(*args, **kwargs)
  File "/opt/odoo/auto/addons/shopinvader_customer_multi_user/services/user_management.py", line 64, in search
    return self._paginate_search(**params)
  File "/opt/odoo/auto/addons/shopinvader/services/service.py", line 141, in _paginate_search
    return {"size": total_count, "data": self._to_json(records, **params)}
  File "/opt/odoo/auto/addons/shopinvader_customer_multi_user/services/user_management.py", line 181, in _to_json
    return records.jsonify(self._json_parser(), **kw)
TypeError: jsonify() got an unexpected keyword argument 'page'
```

HviorForgeFlow replied:
Error related with this merge: https://github.com/shopinvader/odoo-shopinvader/pull/942 @simahawk

simahawk replied:
jsonify accepts kw but not random ones. It happens only when extra params are passed (that's why tests are green). There's something wrong w/ a change I recently made to propagate params across calls. Here we definitely need to sanitize them. I'll check, thanks!

simahawk replied:
> Error related with this merge: #942 @simahawk

yes, exactly, I just have to fix it here for user_management

simahawk replied:
last commit should fix it

simahawk replied:
@HviorForgeFlow can you confirm it's working now?

simahawk committed:
s_customer_multi_user: fix jsonify call

HviorForgeFlow reviewed (approved):
Sure, it is working fine right now

simahawk committed:
shopinvader_customer_multi_user 13.0.2.2.0

simahawk committed:
shopinvader_locomotive 13.0.3.1.1

simahawk commented:
modules' version bumped manually to preserve migration step

/ocabot merge nobump
